rewrite-java/src/main/java/org/openrewrite/java/search/SemanticallyEqual.java

```diff
@@ -1097,13 +1023,16 @@ public J.ParameterizedType visitParameterizedType(J.ParameterizedType type, J j)
         @Override
         public <T extends J> J.Parentheses<T> visitParentheses(J.Parentheses<T> parens, J j) {
             if (isEqual.get()) {
-                if (!(j instanceof J.Parentheses)) {
+                if (j instanceof J.Parentheses) {
+                    J.Parentheses<T> compareTo = (J.Parentheses<T>) j;
+                    this.visit(parens.getTree(), compareTo.getTree());
+                } else if (j instanceof J.ControlParentheses) {
+                    J.ControlParentheses<T> compareTo = (J.ControlParentheses<T>) j;
+                    this.visit(parens.getTree(), compareTo.getTree());
+                } else {
                     isEqual.set(false);
                     return parens;
                 }
```

Comment:
This bit I am a bit unsure about, as I haven't quite understood yet, when `ControlParentheses` gets used and when just `Parentheses`. But for the `InstanceOfPatternMatch` recipe it would be important that these can be compared. Otherwise the recipe would have to do this unpacking itself (which would also be possible).

Reply:
A `J.ControlParentheses` is used in control flow like `if (...)`, `while (...)`, a type cast `((String) object)`, etc.
A `J.Parentheses` is used in logical expressions like `boolean b = null == (obj = null)`.

Reply:
Off the top of my head I do not recall what the difference is... they are similar enough that it does not seem crazy to compare them.